### DISCUSSION TITLE
在windows中图像旋转并保存后，使用paddleocr.PaddleOCR调用det模型检测框变少，并且和直接调用predict_det不一致

### DISCUSSION THREAD

**Discussion by Strepsiades** [PaddlePaddle/PaddleOCR#13511] [Q&A]
#### 问题描述 / Problem Description

这个问题困扰了我一天：我单独用predict_det.py检测图片，检测框很正常。但是使用paddle.PaddleOCR检测一张测试图片，就会少很多检测框。

复现参考下述测试代码，图就不放了，可以照一张横着拍的文字图片。 其中img的形状不随系统中图像旋转而旋转 而img2的形状随图像旋转而旋转。

从脚本调用使用的时img1的方法，从命令行调用使用的是img2的方法。不同输入形状会导致predict_det脚本表现不同。这导致一些非常诡异的检测框丢失现象。

具体来说，如果图像是真实方向，那么检测框就会更多更全。但是如果一个图片本来是放倒的，靠调整变成正的，那么命令行调用predict_det，检测框就正常，脚本paddle.PaddleOcr()调用，就会少检测框。
#### 运行环境 / Runtime Environment
- OS:
- Paddle:
- PaddleOCR:

#### 复现代码 / Reproduction Code
```python
import numpy as np
import cv2
file_path = r"D:\AIdata\training_project\handwrite_ppocrv4\data\20240722204811.png"

# from paddleocr.py line 513
def img_decode(content: bytes):
    np_arr = np.frombuffer(content, dtype=np.uint8)
    return cv2.imdecode(np_arr, cv2.IMREAD_UNCHANGED)

with open(file_path, 'rb') as f:
    img_str = f.read()
    img = img_decode(img_str)
print(img.shape)

# from tools\infer\predict_rec.py line 655
img2 = cv2.imread(file_path)
print(img2.shape)
```

#### 完整报错 / Complete Error Message

(paddle_nlp) PS D:\AIdata\training_project\handwrite_ppocrv4>  d:; cd 'd:\AIdata\training_project\handwrite_ppocrv4'; & 'd:\miniconda\envs\paddle_nlp\python.exe' 'c:\Users\zhaoy\.vscode\extensions\ms-python.debugpy-2024.8.0-win32-x64\bundled\libs\debugpy\adapter/../..\debugpy\launcher' '54864' '--' 'd:\AIdata\training_project\handwrite_ppocrv4\test_opencv.py' 
(1045, 751, 3)
(751, 1045, 3)
(paddle_nlp) PS D:\AIdata\training_project\handwrite_ppocrv4>  d:; cd 'd:\AIdata\training_project\handwrite_ppocrv4'; & 'd:\miniconda\envs\paddle_nlp\python.exe' 'c:\Users\zhaoy\.vscode\extensions\ms-python.debugpy-2024.8.0-win32-x64\bundled\libs\debugpy\adapter/../..\debugpy\launcher' '54883' '--' 'd:\AIdata\training_project\handwrite_ppocrv4\test_opencv.py' 
(1045, 751, 3)
(1045, 751, 3)
(paddle_nlp) PS D:\AIdata\training_project\handwrite_ppocrv4>  d:; cd 'd:\AIdata\training_project\handwrite_ppocrv4'; & 'd:\miniconda\envs\paddle_nlp\python.exe' 'c:\Users\zhaoy\.vscode\extensions\ms-python.debugpy-2024.8.0-win32-x64\bundled\libs\debugpy\adapter/../..\debugpy\launcher' '54898' '--' 'd:\AIdata\training_project\handwrite_ppocrv4\test_opencv.py' 
(1045, 751, 3)
(751, 1045, 3)
#### 可能解决方案 / Possible solutions

把paddleocr.py中的img_decode改成普通的cv2.imread()

**Comment by jingsongliujing** [comment]
只是可视化结果不对吗？可以打印出文本和坐标结果看看完整不？

**Comment by Strepsiades** [comment]
识别的框也变少了，不仅是可视化的问题。比如我把这个网页横着截图，然后在windows中旋转改为正确角度。下方结果1就不能随着系统旋转而旋转，还是横着的。结果2就是正确的方向。结果少识别了OS:三个字。我自己的例子图片方向不对相比于正确可能是15个和20个框的差距。针对文字的方向修正是没用的，因为识别框本身就少了。

![result1](https://github.com/user-attachments/assets/0d6a286c-881b-4be0-bfcc-5024ae0a3170)
![result2](https://github.com/user-attachments/assets/7da9c0ad-6b27-4494-9a64-d2987a812246)


```python
import cv2
from paddleocr import PaddleOCR, draw_ocr

# 初始化OCR模型
ocr = PaddleOCR(use_angle_cls=True, lang='ch', det=True)  # 可根据需要选择语言

# 读取输入图片
image_path = r'D:\AIdata\training_project\handwrite_ppocrv4\data\20240723193639.png'


# 直接调研脚本进行文本检测
result1 = ocr.ocr(image_path, cls=True)
print(result1)

# 先用opencv读取
image = cv2.imread(image_path)
result2 = ocr.ocr(image, cls=True)
print(result2)

# 显示结果1
from PIL import Image 
import numpy as np

def img_decode(content: bytes):
    np_arr = np.frombuffer(content, dtype=np.uint8)
    return cv2.imdecode(np_arr, cv2.IMREAD_UNCHANGED)

with open(image_path, 'rb') as f:
    img_str = f.read()
    img = img_decode(img_str)


result = result1[0]
boxes = [line[0] for line in result]
txts = [line[1][0] for line in result]
scores = [line[1][1] for line in result]
im_show = draw_ocr(img, boxes, txts, scores)
im_show = Image.fromarray(im_show)
im_show.save('result1.jpg')

# 显示结果2
result = result2[0]
boxes = [line[0] for line in result]
txts = [line[1][0] for line in result]
scores = [line[1][1] for line in result]
im_show = draw_ocr(image, boxes, txts, scores)
im_show = Image.fromarray(im_show)
im_show.save('result2.jpg')

```
原始输出是：[2024/07/23 20:01:45] ppocr DEBUG: dt_boxes num : 15, elapsed : 6.175616979598999
[2024/07/23 20:01:46] ppocr DEBUG: cls num  : 15, elapsed : 0.4080312252044678
[2024/07/23 20:01:46] ppocr DEBUG: rec_res num  : 15, elapsed : 0.6476974487304688
[[[[[173.0, 11.0], [206.0, 11.0], [205.0, 331.0], [172.0, 331.0]], ('复现代码／Reproduction Code', 0.959404706954956)], [[[378.0, 15.0], [407.0, 15.0], [407.0, 352.0], [378.0, 352.0]], ('运行环境/RuntimeEnvironment', 0.9958457946777344)], [[[445.0, 15.0], [478.0, 15.0], [478.0, 747.0], [445.0, 747.0]], ('predict_det，检测框就正常，脚本paddle.PaddleOcrO调用，就会少检测框。', 0.9924773573875427)], [[[483.0, 17.检测框丢失现象。', 0.9971337914466858)], [[[571.0, 15.0], [604.0, 15.0], [604.0, 1304.0], [571.0, 1304.0]], ('从脚本调用使用的时img1的方法，从命令行调用使用的是img2的方法。不同输入形状会导致predict_det脚本 表现不同。这导致一些', 0.9928052425384521)], [[[630.0, 14.0], [663.0, 14.0], [659.0, 190.0], [627.0, 189.0]], ('图像旋转而旋转。', 0.9992241263389587)], [[[664.0, 12.0], [696.0, 13.0], [695.0, 705.0], [662.0, 705.0]], ('复现参考下述测试代码，图就不放了，可以照一张横着拍的文字图片。', 0.9989879727363586)], [[[725.0, 15.0], [754.0, 15.0], [754.0, 141.0], [725.0, 141.0]], ('很多检测框。', 0.999688446521759)], [[[757.0, 15.0], [789.0, 15.0], [789.0, 1301.0], [757.0, 1301.0]], ('这个问题困扰了我一天：我单独用predict_det.py检测图片，检测框很正常。但是使用paddle.PaddleOCR检测一张测试图片，就会少', 0.9913232922554016)], [[[814.0, 17.0], [843.0, 17.0], [843.0, 339.0], [814.0, 339.0]], ('问题描述/ProblemDescription', 0.991759181022644)], [[[244.0, 36.0], [272.0, 36.0], [272.0, 183.0], [244.0, 183.0]], ('·PaddleOCR:', 0.9874347448348999)], [[[283.0, 32.0], [311.0, 32.0], [311.0, 144.0], [283.0, 144.0]], ('·Paddle:', 0.9814932346343994)], [[[479.0, 112.0], [511.0, 112.0], [511.0, 1294.0], [479.0, 1294.0]], ('，如果图像是真实方向，那么检测框就会更多更全。但是如果一个图片本来是放倒的，靠调整变成正的，那么命令行调用', 0.9942535758018494)], [[[664.0, 698.0], [696.0, 698.0], [696.0, 1305.0], [664.0, 1305.0]], ('其中img的形状不随系统 中图像旋转而旋转而img2的形状随', 0.999354898929596)]]]
[2024/07/23 20:01:47] ppocr DEBUG: dt_boxes num : 14, elapsed : 0.09145188331604004
[2024/07/23 20:01:47] ppocr DEBUG: cls num  : 14, elapsed : 0.06264233589172363
[2024/07/23 20:01:47] ppocr DEBUG: rec_res num  : 14, elapsed : 0.1908130645751953
[[[[[15.0, 34.0], [335.0, 36.0], [335.0, 63.0], [15.0, 61.0]], ('问题描述/ProblemDescription', 0.9928469657897949)], [[[17.0, 92.0], [1297.0, 90.0], [1297.0, 115.0], [17.0, 116.0]], ('这个问题困扰了我一天：我单独用predict_det.py检测图片，检测框很正常。但是使用paddle.PaddleOCR检测一张测试图片，就会少', 0.9923242926597595)], [[[17.0, 126.0], [134.0, 126.0], [134.0, 149.0], [17.0, 149.0]], ('很多检测框。', 0.9961572289466858)], [[[18.0, 186.0], [1298.0, 186.0], [1298.0, 209.0], [18.0, 209.0]], ('复现参考下述测试代码，图就不放了，可以照一张横看拍的文字图片。其中img的形状不随系统中图像旋转而旋转而img2的形状随', 0.9920384287834167)], [[[17.0, 218.0], [180.0, 218.0], [180.0, 242.0], [17.0, 242.0]], ('图像旋转而旋转。', 0.9975857734680176)], [[[17.0, 277.0], [1298.0, 276.0], [1298.0, 300.0], [17.0, 302.0]], ('从脚本调用 使用的时img1的方法，从命令行调用使用的是img2的方法。不同输入形状会导致predict_det脚本表现不同。这导致一些', 0.998256504535675)], [[[17.0, 311.0], [290.0, 311.0], [290.0, 335.0], [17.0, 335.0]], ('非常诡异的检测框丢失现象。', 0.9865151047706604)], [[[17.0, 370.0], [1287.0, 370.0], [1287.0, 393.0], [17.0, 393.0]], ('具体来说，如果图像是真实方向，那么检测框就会更多更全。但是如果一个图片本来是放倒的，靠调整变成正的，那么命令行调用', 0.9985772371292114)], [[[15.0, 403.0], [745.0, 400.0], [745.0, 427.0], [15.0, 430.0]], ('predict_det，检测框就正常，脚本paddle.PaddleOcrO调用，就会少检测框。', 0.9811303019523621)], [[[17.0, 471.0], [348.0, 471.0], [348.0, 494.0], [17.0, 494.0]], ('运行环境/RuntimeEnvironment', 0.9911704659461975)], [[[36.0, 528.0], [97.0, 528.0], [97.0, 553.0], [36.0, 553.0]], (':SO·', 0.8301111459732056)], [[[36.0, 568.0], [133.0, 568.0], [133.0, 591.0], [36.0, 591.0]], ('·Paddle:', 0.933132529258728)], [[[35.0, 603.0], [178.0, 606.0], [177.0, 631.0], [34.0, 628.0]], ('·PaddleOCR:', 0.9522997140884399)], [[[18.0, 675.0], [323.0, 675.0], [323.0, 698.0], [18.0, 698.0]], ('复现代码/ReproductionCode', 0.9925767183303833)]]]

我看了一下代码，在tools/infer/prediction_system.py第76行dt_boxes, elapse = self.text_detector(img)出来结果就不一样了 
  
```python
def __call__(self, img, cls=True):
        time_dict = {'det': 0, 'rec': 0, 'cls': 0, 'all': 0}

        if img is None:
            logger.debug("no valid image provided")
            return None, None, time_dict

        start = time.time()
        ori_im = img.copy()
        dt_boxes, elapse = self.text_detector(img)
        time_dict['det'] = elapse
```

**Comment by UserWangZz** [comment]
感觉是不是图像读取的问题，可以看看检测前输入模型的图像shape是不是一致的？

**Comment by Strepsiades** [comment]
> 感觉是不是图像读取的问题，可以看看检测前输入模型的图像shape是不是一致的？

是的。目前只要图片旋转过，通过predict_rec.py和通过paddleocr.PaddleOCR调用时，检测前图片shape就不一样，检测后结果也不一样。如果文档是横着的（但是系统里看着是正着的）可能会少一些框。

**Comment by UserWangZz** [comment]
https://blog.csdn.net/sxrkm6s4csx/article/details/117525646
参考一下这个文章，是不是图片里的方向信息导致你读图出现了问题

**Comment by SWHL** [comment]
根据我的经验来看，文本检测框多，最后识别出结果少。一般是因为识别结果时，过滤了置信度<0.5的框。
debug一下没有过滤前的文本检测框和结果，应该是和文本检测框数目是一致的。
